File: FRONT/AmusementPark/src/app/models/parks/park-opening-hours.ts
export interface ParkOpeningHoursTimeRange {
  opensAt: string;
  closesAt: string;
  closesNextDay: boolean;
  lastAdmissionAt?: string | null;
  lastAdmissionNextDay: boolean;
}

export interface ParkOpeningHoursRule {
  id?: string | null;
  startDate: string;
  endDate: string;
  daysOfWeek: string[];
  isClosed: boolean;
  label?: string | null;
  reason?: string | null;
  sortOrder: number;
  timeRanges: ParkOpeningHoursTimeRange[];
}

export interface ParkOpeningHoursDateOverride {
  localDate: string;
  isClosed: boolean;
  label?: string | null;
  reason?: string | null;
  timeRanges: ParkOpeningHoursTimeRange[];
}

export interface ParkOpeningHoursSchedule {
  parkId: string;
  timeZoneId: string;
  sourceUrl?: string | null;
  notes?: string | null;
  lastVerifiedAtUtc?: string | null;
  createdAtUtc?: string | null;
  updatedAtUtc?: string | null;
  regularRules: ParkOpeningHoursRule[];
  dateOverrides: ParkOpeningHoursDateOverride[];
}

export interface ParkOpeningHoursDay {
  localDate: string;
  isClosed: boolean;
  isDefined: boolean;
  sourceKind: string;
  label?: string | null;
  reason?: string | null;
  timeRanges: ParkOpeningHoursTimeRange[];
}

export interface ParkOpeningHoursCalendar {
  parkId: string;
  timeZoneId: string;
  sourceUrl?: string | null;
  notes?: string | null;
  lastVerifiedAtUtc?: string | null;
  updatedAtUtc: string;
  firstDate?: string | null;
  lastDate?: string | null;
  fromDate: string;
  toDate: string;
  days: ParkOpeningHoursDay[];
}

export type ParkOpeningHoursAdminStatus = 'NotConfigured' | 'Expired' | 'NeedsUpdate' | 'UpToDate';
export type ParkOpeningHoursAdminFilter = 'all' | 'configured' | 'notConfigured' | 'upToDate' | 'needsUpdate' | 'expired';

export interface ParkOpeningHoursAdminSummary {
  hasOpeningHours: boolean;
  status: ParkOpeningHoursAdminStatus;
  timeZoneId?: string | null;
  firstDate?: string | null;
  lastDate?: string | null;
  completeUntilDate?: string | null;
  completeForDays?: number | null;
  warningThresholdDays?: number | null;
  lastVerifiedAtUtc?: string | null;
  updatedAtUtc?: string | null;
}

File: API/AmusementPark.Core/Domain/Parks/ParkOpeningHours.cs
namespace AmusementPark.Core.Domain.Parks;

public sealed class ParkOpeningHoursSchedule
{
    public string? Id { get; set; }

    public string ParkId { get; set; } = string.Empty;

    public string TimeZoneId { get; set; } = string.Empty;

    public string? SourceUrl { get; set; }

    public string? Notes { get; set; }

    public DateTime? LastVerifiedAtUtc { get; set; }

    public DateTime CreatedAtUtc { get; set; }

    public DateTime UpdatedAtUtc { get; set; }

    public List<ParkOpeningHoursRule> RegularRules { get; set; } = new();

    public List<ParkOpeningHoursDateOverride> DateOverrides { get; set; } = new();

    public List<ParkOpeningHoursCoverageSegment> CoverageSegments { get; set; } = new();
}

public sealed class ParkOpeningHoursRule
{
    public string? Id { get; set; }

    public DateOnly StartDate { get; set; }

    public DateOnly EndDate { get; set; }

    public List<DayOfWeek> DaysOfWeek { get; set; } = new();

    public bool IsClosed { get; set; }

    public string? Label { get; set; }

    public string? Reason { get; set; }

    public int SortOrder { get; set; }

    public List<ParkOpeningHoursTimeRange> TimeRanges { get; set; } = new();
}

public sealed class ParkOpeningHoursDateOverride
{
    public DateOnly LocalDate { get; set; }

    public bool IsClosed { get; set; }

    public string? Label { get; set; }

    public string? Reason { get; set; }

    public List<ParkOpeningHoursTimeRange> TimeRanges { get; set; } = new();
}

public sealed class ParkOpeningHoursTimeRange
{
    public TimeOnly OpensAt { get; set; }

    public TimeOnly ClosesAt { get; set; }

    public bool ClosesNextDay { get; set; }

    public TimeOnly? LastAdmissionAt { get; set; }

    public bool LastAdmissionNextDay { get; set; }
}

public sealed class ParkOpeningHoursCoverageSegment
{
    public DateOnly StartDate { get; set; }

    public DateOnly EndDate { get; set; }
}
